File: src/main/java/org/apache/prepbuddy/normalizers/MinMaxNormalizer.java
package org.apache.prepbuddy.normalizers;

import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.spark.api.java.JavaDoubleRDD;

import static java.lang.Double.parseDouble;

/**
 * A normalizer which scales the data within the specified range.
 */
public class MinMaxNormalizer implements NormalizationStrategy {
    private final int minRange;
    private final int maxRange;

    private Double minValue;
    private Double maxValue;

    public MinMaxNormalizer(int minRange, int maxRange) {
        this.minRange = minRange;
        this.maxRange = maxRange;
    }

    public MinMaxNormalizer() {
        this(0, 1);
    }

    @Override
    public void prepare(TransformableRDD transformableRDD, int columnIndex) {
        JavaDoubleRDD doubleRDD = transformableRDD.toDoubleRDD(columnIndex);
        maxValue = doubleRDD.max();
        minValue = doubleRDD.min();
    }

    @Override
    public String normalize(String rawValue) {
        double normalizedValue = ((parseDouble(rawValue) - minValue) / (maxValue - minValue)) * (maxRange - minRange) + minRange;
        return String.valueOf(normalizedValue);
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/normalizers/MinMaxNormalizer.scala
package org.apache.datacommons.prepbuddy.normalizers

import org.apache.datacommons.prepbuddy.rdds.TransformableRDD

/**
  * A normalizer which scales the data within the specified range.
  * Default range is (0,1)
  * A' = (A - min(A)) / (max(A) - min(A)) * (D-C) + C
  * where (C,D) is the range and A is the value.
  */
class MinMaxNormalizer(minRange: Int = 0, maxRange: Int = 1) extends NormalizationStrategy {

    private var maxValue: Double = 0
    private var minValue: Double = 0

    override def prepare(transformableRDD: TransformableRDD, columnIndex: Int): Unit = {
        val doubleRDD = transformableRDD.toDoubleRDD(columnIndex)
        maxValue = doubleRDD.max
        minValue = doubleRDD.min
    }

    override def normalize(rawValue: String): String = {
        val normalizedValue: Double = (rawValue.toDouble - minValue) / (maxValue - minValue)
        val normalizedValueInRange = normalizedValue * (maxRange - minRange) + minRange
        normalizedValueInRange.toString
    }
}
